source/slang/slang-parameter-binding.cpp

```diff
@@ -2459,115 +2474,119 @@ RefPtr<ProgramLayout> generateParameterBindings(
         generateParameterBindings(&context, parameter);
     }
 
-    // Determine if there are any global-scope parameters that use `Uniform`
-    // resources, and thus need to get packaged into a constant buffer.
-    //
-    // Note: this doesn't account for GLSL's support for "legacy" uniforms
-    // at global scope, which don't get assigned a CB.
+    // On a CPU target, it's okay to have global scope parameters that use Uniform resources
     bool needDefaultConstantBuffer = false;
-    for( auto& parameterInfo : sharedContext.parameters )
-    {
-        SLANG_RELEASE_ASSERT(parameterInfo->varLayouts.getCount() != 0);
-        auto firstVarLayout = parameterInfo->varLayouts.getFirst();
-
-        // Does the field have any uniform data?
-        if( firstVarLayout->typeLayout->FindResourceInfo(LayoutResourceKind::Uniform) )
-        {
-            needDefaultConstantBuffer = true;
-            diagnoseGlobalUniform(&sharedContext, firstVarLayout->varDecl);
-        }
-    }
-
-    // Next, we want to determine if there are any global-scope parameters
-    // that don't just allocate a whole register space to themselves; these
-    // parameters will need to go into a "default" space, which should always
-    // be the first space we allocate.
-    //
-    // As a starting point, we will definitely need a "default" space if
-    // we are creating a default constant buffer, since it should get
-    // a binding in that "default" space.
-    //
-    bool needDefaultSpace = needDefaultConstantBuffer;
-    if (!needDefaultSpace)
+    if (!_isCPUTarget(targetReq->target))
```

Comment:
I think this `if` is in the wrong place.

In principle we still need/want to scan the global-scope parameters to see if any of them use `Uniform` resources, even on CPU. What we *don't* want to do is make the call to `diagnoseGlobalUniform` and make that condition an error. But the call to  `diagnoseGlobalUniform` as actually a stopgap, because we intend to support global uniforms on all targets, and it just happens that it is not something we've implemented for HLSL/GLSL output.

Reply:
Hmm... actually... I concede the point that even ignoring the diagnostic we emit for HLSL/GLSL targets, you probably don't want this logic triggering *for now* on your CPU target. If we allow the parameter binding logic to wrap up the globals in a constant buffer right now that would be the "correct" layout (in my opinion), but it wouldn't be something your downstream code generation would be able to take advantage of (and in the worst case it would just make the layout you need to work with more complicated).

I *do* think we need to resolve this situation sooner or later by adding the IR pass(es) needed to make global-scope uniforms work. I don't think we are actually too far from having them working anwyay.